service/src/java/org/apache/hive/service/cli/operation/hplsql/HplSqlOperation.java

```diff
@@ -0,0 +1,240 @@
+import org.apache.hadoop.hive.shims.Utils;
+import org.apache.hadoop.security.UserGroupInformation;
+import org.apache.hive.hplsql.Exec;
+import org.apache.hive.hplsql.ResultListener;
+import org.apache.hive.hplsql.executor.Metadata;
+import org.apache.hive.service.cli.FetchOrientation;
+import org.apache.hive.service.cli.HiveSQLException;
+import org.apache.hive.service.cli.OperationState;
+import org.apache.hive.service.cli.OperationType;
+import org.apache.hive.service.cli.RowSet;
+import org.apache.hive.service.cli.RowSetFactory;
+import org.apache.hive.service.cli.TableSchema;
+import org.apache.hive.service.cli.operation.ExecuteStatementOperation;
+import org.apache.hive.service.cli.session.HiveSession;
+import org.apache.hive.service.server.ThreadWithGarbageCleanup;
+
+public class HplSqlOperation extends ExecuteStatementOperation implements ResultListener {
+  private final Exec exec;
+  private final boolean runInBackground;
+  private RowSet rowSet;
+  private TableSchema schema;
+
+  public HplSqlOperation(HiveSession parentSession, String statement, Map<String, String> confOverlay, boolean runInBackground, Exec exec) {
+    super(parentSession, statement, confOverlay, runInBackground, false);
+    this.exec = exec;
+    this.runInBackground = runInBackground;
+    exec.setResultListener(this);
+  }
+
+  @Override
+  protected void runInternal() throws HiveSQLException {
+    setState(OperationState.PENDING);
+    if (!runInBackground) {
+      interpret();
+    } else {
+      Runnable work = new BackgroundWork(getCurrentUGI(), parentSession.getSessionHive(), SessionState.get());
+      try {
+        // This submit blocks if no background threads are available to run this operation
+        Future<?> backgroundHandle = getParentSession().submitBackgroundOperation(work);
+        setBackgroundHandle(backgroundHandle);
+      } catch (RejectedExecutionException rejected) {
+        setState(OperationState.ERROR);
+        throw new HiveSQLException("The background threadpool cannot accept" +
+                " new task for execution, please retry the operation", rejected);
+      }
+    }
+  }
+
+  private void interpret() throws HiveSQLException {
+    try {
+      OperationState opState = getStatus().getState();
+      // Operation may have been cancelled by another thread
+      if (opState.isTerminal()) {
+        log.info("Not running the query. Operation is already in terminal state: " + opState
+                + ", perhaps cancelled due to query timeout or by another thread.");
+        return;
+      }
+      setState(OperationState.RUNNING);
+      int code = exec.run(new String[]{"-e", statement});
+      if (code != 0) {
+        throw new HiveSQLException("HPL/SQL returned " + code);
+      }
+      setState(OperationState.FINISHED);
+    } catch (Throwable e) {
+      if (getStatus().getState().isTerminal()) {
+        log.warn("Ignore exception in terminal state: {}", getStatus().getState(), e);
```

Comment:
I think this would not log `e` without a matching `{}`

Reply:
I tried it, and it still works this way. The same way is used at other places in the code.